src/finalizer/utils/helios.ts

```
@@ -0,0 +1,927 @@
+          l1TxHash: l1Event.transactionHash,
+          nonce: nonce.toString(),
+          storageSlot: expectedStorageSlot,
+          verifiedOnL2TxHash: verifiedEvent.transactionHash,
+          verifiedHead: verifiedEvent.args.head.toString(),
+        });
+      } else {
+        // Not verified and not executed -> Needs Proof and Execution
+        pendingMessages.push({
+          l1Event: l1Event,
+          status: "NeedsProofAndExecution",
+          // verifiedHead is undefined here
+        });
+      }
+    }
+    // If isExecuted is true, the message is fully finalized, do nothing.
+  }
+
+  logger.debug({
+    at: `Finalizer#identifyPendingHeliosMessages:${l2ChainId}`,
+    message: "Finished identifying pending messages.",
+    totalL1StoredCallData: relevantStoredCallDataEvents.length,
+    totalL2VerifiedSlots: verifiedSlotsMap.size,
+    totalL2RelayedNonces: relayedNonces.size,
+    pendingMessagesCount: pendingMessages.length,
+    needsProofCount: pendingMessages.filter((m) => m.status === "NeedsProofAndExecution").length,
+    needsExecutionOnlyCount: pendingMessages.filter((m) => m.status === "NeedsExecutionOnly").length,
+  });
+
+  return pendingMessages;
+}
+
+/** Query and Filter L1 Events */
+async function getRelevantL1Events(
+  logger: winston.Logger,
+  hubPoolClient: HubPoolClient,
+  l1SpokePoolClient: SpokePoolClient,
+  l1ChainId: number,
+  l2ChainId: number,
+  l2SpokePoolAddress: string
+): Promise<StoredCallDataEvent[] | null> {
+  const l1Provider = hubPoolClient.hubPool.provider;
+
+  const hubPoolStoreInfo = CONTRACT_ADDRESSES[l1ChainId]?.hubPoolStore;
+  if (!hubPoolStoreInfo?.address || !hubPoolStoreInfo.abi) {
+    logger.error({
+      at: `Finalizer#heliosL1toL2Finalizer:getAndFilterL1Events:${l2ChainId}`,
+      message: `HubPoolStore contract address or ABI not found for L1 chain ${l1ChainId}.`,
+    });
+    return null;
```

you should throw here instead of returning null. We should always have the ABI/address available to us otherwise its the coder's error